content/post/2024-03-12-mocking/index.Rmd

```diff
@@ -0,0 +1,143 @@
+```
+
+Find the example repository on [GitHub](https://github.com/maelle/mockexample).
+
+The mockexample package has a function called `is_encoding_a_pain()`:
+
+``` r
+is_encoding_a_pain <- function(){
+  if (Sys.info()[["sysname"]] == "Windows"){
+    return(TRUE)
+  }
+
+  FALSE
+}
+```
+
+The function returns `TRUE` on Windows, `FALSE` on other operating systems.
+
+To test it no matter the operating system, we need to make our tests believe they are run on a given operating system.
+We can in this day and age achieve this using mocking tools from testthat itself!
+Here's our test file:
+
+``` r
+test_that("is_encoding_a_pain() works on Linux", {
+  local_mocked_bindings(Sys.info = function(...) c(sysname = "Linux"))
+  expect_false(is_encoding_a_pain())
+})
+
+test_that("is_encoding_a_pain() works on Windows", {
+  local_mocked_bindings(Sys.info = function(...) c(sysname = "Windows"))
+  expect_true(is_encoding_a_pain())
+})
+```
+
+The magic happens in `local_mocked_bindings()` calls such as `local_mocked_bindings(Sys.info = function(...) c(sysname = "Windows"))`, that indicates what we want the mocked function, `Sys.info()`, to return in this environment.
+Using testthat's own mocking feels good for three reasons: not needing to add another dependency (and to remember what the mocking packages are called :sweat_smile:), knowing this mocking implementation is [considered better](https://www.tidyverse.org/blog/2023/10/testthat-3-2-0/#mocking) by testthat authors, and getting a [withr vibe](https://withr.r-lib.org/) from the "local\_" aspect.
+
+Now, there's a catch here because `Sys.info()` is a base R function.
+We had to add this line in the `R/encoding.R` script:
+
+``` r
+Sys.info <- NULL
+```
+
+This case is documented, as other cases (namespaced calls in particular), in [`testthat::local_mocked_bindings()` manual page](https://testthat.r-lib.org/reference/local_mocked_bindings.html#use).
+
+TODO: example from keyring?
+
+## Revisiting the escape hatch example
+
+Even the example in a more recent post, on [escape hatches](/2023/01/23/code-switch-escape-hatch-test/), can be easily re-written to use mocking: we'd use `local_mocked_bindings(is_internet_down = function(...) TRUE)`.
+
+Mocking, and its nowadays improved support in testthat, does not completely throw the escape hatch pattern out of the window, but it's still good to keep mocking in mind.
+That post actually also mentioned mocking so it's all good.
+:innocent:
+
+## Conclusion
+
+In this post, we corrected the [someone who was wrong on the internet](https://xkcd.com/386/) years ago.
+In summary, nowadays you can use mocking tools from testthat itself, when you decide mocking is more a help than a hindrance in your testing suite, and when you do not need to turn to specific packages such as [httptest2](https://enpiar.com/httptest2/), [vcr](https://docs.ropensci.org/vcr/) (for HTTP calls), [dittodb](https://docs.ropensci.org/dittodb/) (for database calls).
```

Comment:
These special cases are part of the 2019 post but not this one, so they are a little bit of a surprise in a _summary_  for people who only read this post.